File: Sources/LaTeXUnicode/Converter.swift
/// Converts a LaTeX fragment to inline Unicode.
///
/// Accepts a whole fragment, not just a single command: literal text passes
/// through untouched, so `convert("x^2 + \\alpha")` yields `x² + α`.
///
/// The result is all-or-nothing. If any part of the input has no faithful
/// Unicode form, the whole call returns `.unsupported` — a half-converted
/// string is never produced.
public func convert(_ latex: String) -> ConversionResult {
    guard !latex.isEmpty else {
        return .unsupported(reason: "There is nothing to convert.")
    }
    do {
        let rendered = try Renderer.render(try Tokenizer.tokenize(latex))
        guard !rendered.fallbacks.isEmpty else {
            return .converted(rendered.text)
        }
        return .fallback(rendered.text, reason: deduplicated(rendered.fallbacks).joined(separator: " "))
    } catch let error as UnsupportedInput {
        return .unsupported(reason: error.reason)
    } catch let error as TokenizerError {
        return .unsupported(reason: error.reason)
    } catch {
        return .unsupported(reason: "The input could not be parsed.")
    }
}

// MARK: - Internals

private struct Rendered {
    var text = ""
    var fallbacks: [String] = []

    mutating func append(_ other: Rendered) {
        text += other.text
        fallbacks += other.fallbacks
    }
}

private struct UnsupportedInput: Error {
    let reason: String
}

private enum ScriptKind {
    case sup, sub

    var name: String { self == .sup ? "superscript" : "subscript" }
    var marker: Character { self == .sup ? "^" : "_" }
    var table: [Character: Character] {
        self == .sup ? ScriptTables.superscripts : ScriptTables.subscripts
    }
}

private struct Renderer {
    let tokens: [Token]
    var index = 0
    var out = Rendered()

    static func render(_ tokens: [Token]) throws -> Rendered {
        var renderer = Renderer(tokens: tokens)
        try renderer.run()
        return renderer.out
    }

    private mutating func run() throws {
        while index < tokens.count {
            let token = tokens[index]
            index += 1
            switch token {
            case .character(let c):     out.text.append(c)
            case .group(let inner):     out.append(try Renderer.render(inner))
            case .sup:                  try applyScript(.sup)
            case .sub:                  try applyScript(.sub)
            case .controlSymbol(let c): try emit(controlSymbol: c)
            case .command(let name):    try emit(command: name)
            }
        }
    }

    /// The next argument: a braced group's contents, or a single token.
    /// `\frac12` and `\frac{1}{2}` therefore behave the same, as in LaTeX.
    ///
    /// Whitespace before an argument is skipped, so `\mathbb R` and `\frac 1 2`
    /// work. This does not contradict the tokenizer's rule that a command's
    /// terminator is preserved: that rule is about symbol commands, which take
    /// no argument and so never reach this method.
    private mutating func nextUnit() -> [Token]? {
        while index < tokens.count,
              case .character(let c) = tokens[index], c.isWhitespace {
            index += 1
        }
        guard index < tokens.count else { return nil }
        let token = tokens[index]
        index += 1
        if case .group(let inner) = token { return inner }
        return [token]
    }

    // MARK: Scripts

    private mutating func applyScript(_ kind: ScriptKind) throws {
        guard let unit = nextUnit() else {
            throw UnsupportedInput(reason: "`\(kind.marker)` has nothing after it to convert.")
        }
        guard !unit.isEmpty else {
            throw UnsupportedInput(reason: "The \(kind.name) is empty.")
        }
        guard !containsScript(unit) else {
            throw UnsupportedInput(
                reason: "Nested superscripts and subscripts have no Unicode form.")
        }

        let inner = try Renderer.render(unit)
        for character in inner.text {
            guard let mapped = kind.table[character] else {
                throw UnsupportedInput(
                    reason: "There is no Unicode \(kind.name) for \u{201C}\(character)\u{201D}.")
            }
            out.text.append(mapped)
        }
        out.fallbacks += inner.fallbacks
    }

    // MARK: Commands

    private mutating func emit(command name: String) throws {
        switch name {
        case "frac", "dfrac", "tfrac":
            return try emitFraction(name)
        case "sqrt":
            return try emitRoot()
        case "binom", "dbinom", "tbinom":
            return try emitBinomial(name)
        case "text", "textrm", "mathrm", "operatorname":
            return try emitVerbatimGroup(name)
        case "mathbb":
            return try emitBlackboardBold()
        case "left", "right":
            // Delimiter sizing only. The delimiter itself is the next token.
            return
        case "begin", "end":
            return try rejectEnvironment()
        default:
            break
        }

        if let text = TextOperators.operators[name] {
            out.text += text
            return
        }
        if let reason = UnsupportedCommands.reasons[name] {
            throw UnsupportedInput(reason: reason)
        }
        if let symbol = SymbolTable.symbols[name] {
            out.text += symbol
            return
        }
        throw UnsupportedInput(reason: "Unknown command \\\(name).")
    }

    private mutating func emit(controlSymbol character: Character) throws {
        if let literal = UnsupportedCommands.escapedLiterals[character] {
            out.text.append(literal)
            return
        }
        if let reason = UnsupportedCommands.symbolReasons[character] {
            throw UnsupportedInput(reason: reason)
        }
        throw UnsupportedInput(reason: "Unknown command \\\(character).")
    }

    // MARK: Fallback constructions

    private mutating func emitFraction(_ name: String) throws {
        guard let numerator = nextUnit(), let denominator = nextUnit() else {
            throw UnsupportedInput(
                reason: "\\\(name) needs two arguments, for example \\\(name){a}{b}.")
        }
        let top = try Renderer.render(numerator)
        let bottom = try Renderer.render(denominator)
        guard !top.text.isEmpty, !bottom.text.isEmpty else {
            throw UnsupportedInput(reason: "\\\(name) has an empty argument.")
        }
        // A fraction whose own arguments already needed an approximation
        // cannot be represented exactly, so those go straight to the linear
        // form rather than dressing up a fallback as a real fraction.
        if top.fallbacks.isEmpty, bottom.fallbacks.isEmpty {
            // Best: a single precomposed character.
            if let exact = ScriptTables.vulgarFractions["\(top.text)/\(bottom.text)"] {
                out.text += exact
                return
            }
            // Next best: superscript numerator, fraction slash, subscript
            // denominator. This composes any fraction whose digits and letters
            // both have script forms, so 10/17 works even though no single
            // character for it exists.
            if worthComposing(numerator: top.text, denominator: bottom.text),
               let composed = composedFraction(numerator: top.text, denominator: bottom.text) {
                out.text += composed
                return
            }
        }

        out.text += mathematical(parenthesised(top.text))
            + String(divisionSlash)
            + mathematical(parenthesised(bottom.text))
        out.fallbacks += top.fallbacks + bottom.fallbacks
        out.fallbacks.append(
            "A fraction cannot be stacked in plain text, so it was written on one line with a slash.")
    }

    /// Whether a composed fraction would still be readable.
    ///
    /// Script characters are small, and superscript `x` against subscript `x`
    /// is close to indistinguishable at text size. Digits stay legible at any
    /// length, so `10/17` composes; anything longer than a couple of characters
    /// with letters in it reads better on one line.
    private func worthComposing(numerator: String, denominator: String) -> Bool {
        let bothNumeric = numerator.allSatisfy(\.isNumber) && denominator.allSatisfy(\.isNumber)
        let bothShort = numerator.count <= 2 && denominator.count <= 2
        return bothNumeric || bothShort
    }

    /// Builds a diagonal fraction out of existing script characters.
    ///
    /// Returns nil when any character lacks the form it needs — the subscript
    /// alphabet is missing `b c d f g q w y z`, so `\frac{a}{b}` cannot be
    /// composed and falls back to `a/b` instead.
    private func composedFraction(numerator: String, denominator: String) -> String? {
        var result = ""
        for character in numerator {
            guard let raised = ScriptTables.superscripts[character] else { return nil }
            result.append(raised)
        }
        result.append(ScriptTables.fractionSlash)
        for character in denominator {
            guard let lowered = ScriptTables.subscripts[character] else { return nil }
            result.append(lowered)
        }
        return result
    }

    private mutating func emitRoot() throws {
        var degree = 2
        var degreeText = "2"

        // Optional [n]. `[` and `]` are ordinary characters everywhere else,
        // so the bracket group is only recognised here, immediately after \sqrt.
        if index < tokens.count, tokens[index] == .character("[") {
            var scan = index + 1
            var inner: [Token] = []
            while scan < tokens.count, tokens[scan] != .character("]") {
                inner.append(tokens[scan])
                scan += 1
            }
            guard scan < tokens.count else {
                throw UnsupportedInput(reason: "\\sqrt has a `[` that is never closed.")
            }
            index = scan + 1
            degreeText = try Renderer.render(inner).text

            // An index that is missing or not a root produced honest-looking
            // nonsense: `\\sqrt[]{8}` came out as `8^(1/)` and `\\sqrt[0]{8}` as
            // `8^(1/0)`. A letter index is fine and common, `\\sqrt[n]{x}` reads
            // perfectly well as a fractional power; only a number below two is
            // not a root at all.
            guard !degreeText.isEmpty else {
                throw UnsupportedInput(
                    reason: "\\sqrt has an empty index, for example \\sqrt[3]{8}.")
            }
            if let written = Int(degreeText), written < 2 {
                throw UnsupportedInput(
                    reason: "A root needs an index of at least 2, so \\sqrt[\(written)] is not one.")
            }
            degree = Int(degreeText) ?? -1
        }

        guard let argument = nextUnit() else {
            throw UnsupportedInput(reason: "\\sqrt needs an argument, for example \\sqrt{2}.")
        }
        let radicand = try Renderer.render(argument)
        guard !radicand.text.isEmpty else {
            throw UnsupportedInput(reason: "\\sqrt has an empty argument.")
        }
        out.fallbacks += radicand.fallbacks

        if let radical = ScriptTables.radicals[degree] {
            out.text += String(radical) + mathematical(parenthesised(radicand.text))
            out.fallbacks.append(
                "A radical sign cannot extend over what is under it, so the root was written as \(radical)(\u{2026}).")
        } else {
            out.text += mathematical(parenthesised(radicand.text)) + "^(1/" + degreeText + ")"
            out.fallbacks.append(
                "There is no Unicode radical sign for an index of \(degreeText), so the root was written as a fractional power.")
        }
    }

    private mutating func emitBinomial(_ name: String) throws {
        guard let upper = nextUnit(), let lower = nextUnit() else {
            throw UnsupportedInput(
                reason: "\\\(name) needs two arguments, for example \\\(name){n}{k}.")
        }
        let n = try Renderer.render(upper)
        let k = try Renderer.render(lower)
        guard !n.text.isEmpty, !k.text.isEmpty else {
            throw UnsupportedInput(reason: "\\\(name) has an empty argument.")
        }
        out.text += "C(\(mathematical(n.text)), \(mathematical(k.text)))"
        out.fallbacks += n.fallbacks + k.fallbacks
        out.fallbacks.append(
            "A binomial coefficient cannot be stacked in plain text, so it was written as C(n, k).")
    }

    // MARK: Passthrough constructions

    private mutating func emitVerbatimGroup(_ name: String) throws {
        guard let unit = nextUnit() else {
            throw UnsupportedInput(reason: "\\\(name) needs an argument.")
        }
        out.append(try Renderer.render(unit))
    }

    private mutating func emitBlackboardBold() throws {
        guard let unit = nextUnit() else {
            throw UnsupportedInput(reason: "\\mathbb needs a letter, for example \\mathbb{R}.")
        }
        let letter = try Renderer.render(unit).text
        guard letter.count == 1, let symbol = SymbolTable.symbols[letter] else {
            throw UnsupportedInput(
                reason: "There is no double-struck Unicode letter for \u{201C}\(letter)\u{201D}.")
        }
        out.text += symbol
    }

    private mutating func rejectEnvironment() throws {
        guard let unit = nextUnit(), let name = try? Renderer.render(unit).text, !name.isEmpty else {
            throw UnsupportedInput(
                reason: "LaTeX environments need two-dimensional layout, which has no inline Unicode form.")
        }
        throw UnsupportedInput(
            reason: "The \(name) environment needs two-dimensional layout, which has no inline Unicode form.")
    }
}

// MARK: - Helpers

/// U+2215, the mathematical division operator, rather than the ASCII solidus.
private let divisionSlash: Character = "\u{2215}"

/// Replaces the ASCII hyphen with U+2212 MINUS SIGN.
///
/// A hyphen is not a minus: it is shorter, sits lower, and reads as a word
/// break. Only linear maths goes through here — composed fractions and scripts
/// use their own precomposed minus characters, and `\text{}` keeps its hyphens
/// so "well-known" is not mangled.
private func mathematical(_ text: String) -> String {
    String(text.map { $0 == "-" ? "\u{2212}" : $0 })
}

/// Wraps in parentheses unless it is a single character, so `\frac{1}{2}`
/// gives `½` while `\frac{x+1}{y-2}` gives `(x+1)∕(y−2)`.
private func parenthesised(_ text: String) -> String {
    text.count == 1 ? text : "(" + text + ")"
}

private func containsScript(_ tokens: [Token]) -> Bool {
    tokens.contains { token in
        switch token {
        case .sup, .sub:        return true
        case .group(let inner): return containsScript(inner)
        default:                return false
        }
    }
}

private func deduplicated(_ items: [String]) -> [String] {
    var seen = Set<String>()
    return items.filter { seen.insert($0).inserted }
}

File: windows/LaTeXSquiggly.Core/Engine/Converter.cs
using System.Text;

namespace LaTeXSquiggly.Core.Engine;

/// <summary>
/// Converts a LaTeX fragment to inline Unicode.
///
/// Accepts a whole fragment, not just a single command: literal text passes
/// through untouched, so Convert("x^2 + \\alpha") yields "x² + α".
///
/// The result is all-or-nothing. If any part of the input has no faithful
/// Unicode form the whole call returns Unsupported: a half-converted string is
/// never produced.
/// </summary>
public static class Converter
{
    public static ConversionResult Convert(string latex)
    {
        if (latex.Length == 0)
        {
            return ConversionResult.Unsupported("There is nothing to convert.");
        }
        try
        {
            var rendered = Renderer.Render(Tokenizer.Tokenize(latex));
            if (rendered.Fallbacks.Count == 0)
            {
                return ConversionResult.Converted(rendered.Text.ToString());
            }
            return ConversionResult.Fallback(
                rendered.Text.ToString(), string.Join(" ", Deduplicated(rendered.Fallbacks)));
        }
        catch (UnsupportedInputException error)
        {
            return ConversionResult.Unsupported(error.Message);
        }
        catch (TokenizerException error)
        {
            return ConversionResult.Unsupported(error.Message);
        }
    }

    private static List<string> Deduplicated(IEnumerable<string> items)
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var item in items)
        {
            if (seen.Add(item)) result.Add(item);
        }
        return result;
    }
}

internal sealed class UnsupportedInputException : Exception
{
    public UnsupportedInputException(string reason) : base(reason) { }
}

internal sealed class Rendered
{
    public StringBuilder Text { get; } = new();
    public List<string> Fallbacks { get; } = new();

    public void Append(Rendered other)
    {
        Text.Append(other.Text);
        Fallbacks.AddRange(other.Fallbacks);
    }
}

internal enum ScriptKind { Sup, Sub }

internal static class ScriptKindExtensions
{
    public static string Name(this ScriptKind kind) =>
        kind == ScriptKind.Sup ? "superscript" : "subscript";

    public static char Marker(this ScriptKind kind) => kind == ScriptKind.Sup ? '^' : '_';

    public static IReadOnlyDictionary<char, char> Table(this ScriptKind kind) =>
        kind == ScriptKind.Sup ? ScriptTables.Superscripts : ScriptTables.Subscripts;
}

internal sealed class Renderer
{
    /// <summary>U+2215, the division operator, rather than the ASCII solidus.</summary>
    private const char DivisionSlash = '∕';

    private readonly IReadOnlyList<Token> _tokens;
    private int _index;
    private readonly Rendered _out = new();

    private Renderer(IReadOnlyList<Token> tokens) => _tokens = tokens;

    public static Rendered Render(IReadOnlyList<Token> tokens)
    {
        var renderer = new Renderer(tokens);
        renderer.Run();
        return renderer._out;
    }

    private void Run()
    {
        while (_index < _tokens.Count)
        {
            var token = _tokens[_index];
            _index += 1;
            switch (token)
            {
                case Token.Character character: _out.Text.Append(character.Value); break;
                case Token.Group group: _out.Append(Render(group.Inner)); break;
                case Token.Sup: ApplyScript(ScriptKind.Sup); break;
                case Token.Sub: ApplyScript(ScriptKind.Sub); break;
                case Token.ControlSymbol symbol: EmitControlSymbol(symbol.Symbol); break;
                case Token.Command command: EmitCommand(command.Name); break;
            }
        }
    }

    /// <summary>
    /// The next argument: a braced group's contents, or a single token, so
    /// <c>\frac12</c> and <c>\frac{1}{2}</c> behave the same, as in LaTeX.
    ///
    /// Whitespace before an argument is skipped, so <c>\mathbb R</c> works. This
    /// does not contradict the tokenizer's rule that a command's terminator is
    /// preserved: that rule is about symbol commands, which take no argument and
    /// so never reach here.
    /// </summary>
    private IReadOnlyList<Token>? NextUnit()
    {
        while (_index < _tokens.Count
               && _tokens[_index] is Token.Character white
               && char.IsWhiteSpace(white.Value))
        {
            _index += 1;
        }
        if (_index >= _tokens.Count) return null;

        var token = _tokens[_index];
        _index += 1;
        if (token is Token.Group group) return group.Inner;
        return new[] { token };
    }

    // MARK: Scripts

    private void ApplyScript(ScriptKind kind)
    {
        var unit = NextUnit();
        if (unit is null)
        {
            throw new UnsupportedInputException(
                $"`{kind.Marker()}` has nothing after it to convert.");
        }
        if (unit.Count == 0)
        {
            throw new UnsupportedInputException($"The {kind.Name()} is empty.");
        }
        if (ContainsScript(unit))
        {
            throw new UnsupportedInputException(
                "Nested superscripts and subscripts have no Unicode form.");
        }

        var inner = Render(unit);
        // Whole characters, not UTF-16 halves: a double-struck letter is one
        // character to the user and to Swift, so a refusal has to name it as one.
        foreach (var character in Core.Text.Characters(inner.Text.ToString()))
        {
            if (character.Length != 1 || !kind.Table().TryGetValue(character[0], out var mapped))
            {
                throw new UnsupportedInputException(
                    $"There is no Unicode {kind.Name()} for “{character}”.");
            }
            _out.Text.Append(mapped);
        }
        _out.Fallbacks.AddRange(inner.Fallbacks);
    }

    // MARK: Commands

    private void EmitCommand(string name)
    {
        switch (name)
        {
            case "frac" or "dfrac" or "tfrac": EmitFraction(name); return;
            case "sqrt": EmitRoot(); return;
            case "binom" or "dbinom" or "tbinom": EmitBinomial(name); return;
            case "text" or "textrm" or "mathrm" or "operatorname": EmitVerbatimGroup(name); return;
            case "mathbb": EmitBlackboardBold(); return;
            // Delimiter sizing only. The delimiter itself is the next token.
            case "left" or "right": return;
            case "begin" or "end": RejectEnvironment(); return;
        }

        if (TextOperators.Operators.TryGetValue(name, out var operatorText))
        {
            _out.Text.Append(operatorText);
            return;
        }
        if (UnsupportedCommands.Reasons.TryGetValue(name, out var reason))
        {
            throw new UnsupportedInputException(reason);
        }
        if (SymbolTable.Symbols.TryGetValue(name, out var symbol))
        {
            _out.Text.Append(symbol);
            return;
        }
        throw new UnsupportedInputException($"Unknown command \\{name}.");
    }

    private void EmitControlSymbol(char character)
    {
        if (UnsupportedCommands.EscapedLiterals.TryGetValue(character, out var literal))
        {
            _out.Text.Append(literal);
            return;
        }
        if (UnsupportedCommands.SymbolReasons.TryGetValue(character, out var reason))
        {
            throw new UnsupportedInputException(reason);
        }
        throw new UnsupportedInputException($"Unknown command \\{character}.");
    }

    // MARK: Fallback constructions

    private void EmitFraction(string name)
    {
        var numerator = NextUnit();
        var denominator = NextUnit();
        if (numerator is null || denominator is null)
        {
            throw new UnsupportedInputException(
                $"\\{name} needs two arguments, for example \\{name}{{a}}{{b}}.");
        }
        var top = Render(numerator);
        var bottom = Render(denominator);
        var topText = top.Text.ToString();
        var bottomText = bottom.Text.ToString();
        if (topText.Length == 0 || bottomText.Length == 0)
        {
            throw new UnsupportedInputException($"\\{name} has an empty argument.");
        }

        // A fraction whose own arguments already needed an approximation cannot
        // be represented exactly, so those go straight to the linear form rather
        // than dressing up a fallback as a real fraction.
        if (top.Fallbacks.Count == 0 && bottom.Fallbacks.Count == 0)
        {
            // Best: a single precomposed character.
            if (ScriptTables.VulgarFractions.TryGetValue($"{topText}/{bottomText}", out var exact))
            {
                _out.Text.Append(exact);
                return;
            }
            // Next best: superscript numerator, fraction slash, subscript
            // denominator, which composes anything whose parts both have script
            // forms, so 10/17 works though no single character for it exists.
            if (WorthComposing(topText, bottomText))
            {
                var composed = ComposedFraction(topText, bottomText);
                if (composed is not null)
                {
                    _out.Text.Append(composed);
                    return;
                }
            }
        }

        _out.Text.Append(Mathematical(Parenthesised(topText)))
                 .Append(DivisionSlash)
                 .Append(Mathematical(Parenthesised(bottomText)));
        _out.Fallbacks.AddRange(top.Fallbacks);
        _out.Fallbacks.AddRange(bottom.Fallbacks);
        _out.Fallbacks.Add(
            "A fraction cannot be stacked in plain text, so it was written on one line with a slash.");
    }

    /// <summary>
    /// Whether a composed fraction would still be readable. Script characters
    /// are small, and superscript x against subscript x is close to
    /// indistinguishable at text size. Digits stay legible at any length.
    /// </summary>
    private static bool WorthComposing(string numerator, string denominator)
    {
        var bothNumeric = numerator.All(char.IsNumber) && denominator.All(char.IsNumber);
        var bothShort = Core.Text.CharacterCount(numerator) <= 2
                        && Core.Text.CharacterCount(denominator) <= 2;
        return bothNumeric || bothShort;
    }

    /// <summary>
    /// Builds a diagonal fraction out of existing script characters. Null when
    /// any character lacks the form it needs: the subscript alphabet is missing
    /// b c d f g q w y z, so <c>\frac{a}{b}</c> falls back to a/b instead.
    /// </summary>
    private static string? ComposedFraction(string numerator, string denominator)
    {
        var result = new StringBuilder();
        foreach (var character in numerator)
        {
            if (!ScriptTables.Superscripts.TryGetValue(character, out var raised)) return null;
            result.Append(raised);
        }
        result.Append(ScriptTables.FractionSlash);
        foreach (var character in denominator)
        {
            if (!ScriptTables.Subscripts.TryGetValue(character, out var lowered)) return null;
            result.Append(lowered);
        }
        return result.ToString();
    }

    private void EmitRoot()
    {
        var degree = 2;
        var degreeText = "2";

        // Optional [n]. `[` and `]` are ordinary characters everywhere else, so
        // the bracket group is only recognised here, immediately after \sqrt.
        if (_index < _tokens.Count
            && _tokens[_index] is Token.Character { Value: '[' })
        {
            var scan = _index + 1;
            var inner = new List<Token>();
            while (scan < _tokens.Count && _tokens[scan] is not Token.Character { Value: ']' })
            {
                inner.Add(_tokens[scan]);
                scan += 1;
            }
            if (scan >= _tokens.Count)
            {
                throw new UnsupportedInputException("\\sqrt has a `[` that is never closed.");
            }
            _index = scan + 1;
            degreeText = Render(inner).Text.ToString();

            // An index that is missing or not a root produced honest-looking
            // nonsense: \sqrt[]{8} came out as 8^(1/) and \sqrt[0]{8} as
            // 8^(1/0). A letter index is fine and common, \sqrt[n]{x} reads
            // perfectly well as a fractional power; only a number below two is
            // not a root at all.
            if (degreeText.Length == 0)
            {
                throw new UnsupportedInputException(
                    "\\sqrt has an empty index, for example \\sqrt[3]{8}.");
            }
            if (int.TryParse(degreeText, out var written) && written < 2)
            {
                throw new UnsupportedInputException(
                    $"A root needs an index of at least 2, so \\sqrt[{written}] is not one.");
            }
            degree = int.TryParse(degreeText, out var parsed) ? parsed : -1;
        }

        var argument = NextUnit();
        if (argument is null)
        {
            throw new UnsupportedInputException("\\sqrt needs an argument, for example \\sqrt{2}.");
        }
        var radicand = Render(argument);
        var radicandText = radicand.Text.ToString();
        if (radicandText.Length == 0)
        {
            throw new UnsupportedInputException("\\sqrt has an empty argument.");
        }
        _out.Fallbacks.AddRange(radicand.Fallbacks);

        if (ScriptTables.Radicals.TryGetValue(degree, out var radical))
        {
            _out.Text.Append(radical).Append(Mathematical(Parenthesised(radicandText)));
            _out.Fallbacks.Add(
                $"A radical sign cannot extend over what is under it, so the root was written as {radical}(…).");
        }
        else
        {
            _out.Text.Append(Mathematical(Parenthesised(radicandText)))
                     .Append("^(1/").Append(degreeText).Append(')');
            _out.Fallbacks.Add(
                $"There is no Unicode radical sign for an index of {degreeText}, so the root was written as a fractional power.");
        }
    }

    private void EmitBinomial(string name)
    {
        var upper = NextUnit();
        var lower = NextUnit();
        if (upper is null || lower is null)
        {
            throw new UnsupportedInputException(
                $"\\{name} needs two arguments, for example \\{name}{{n}}{{k}}.");
        }
        var n = Render(upper);
        var k = Render(lower);
        var nText = n.Text.ToString();
        var kText = k.Text.ToString();
        if (nText.Length == 0 || kText.Length == 0)
        {
            throw new UnsupportedInputException($"\\{name} has an empty argument.");
        }
        _out.Text.Append($"C({Mathematical(nText)}, {Mathematical(kText)})");
        _out.Fallbacks.AddRange(n.Fallbacks);
        _out.Fallbacks.AddRange(k.Fallbacks);
        _out.Fallbacks.Add(
            "A binomial coefficient cannot be stacked in plain text, so it was written as C(n, k).");
    }

    // MARK: Passthrough constructions

    private void EmitVerbatimGroup(string name)
    {
        var unit = NextUnit();
        if (unit is null)
        {
            throw new UnsupportedInputException($"\\{name} needs an argument.");
        }
        _out.Append(Render(unit));
    }

    private void EmitBlackboardBold()
    {
        var unit = NextUnit();
        if (unit is null)
        {
            throw new UnsupportedInputException(
                "\\mathbb needs a letter, for example \\mathbb{R}.");
        }
        var letter = Render(unit).Text.ToString();
        if (Core.Text.CharacterCount(letter) != 1
            || !SymbolTable.Symbols.TryGetValue(letter, out var symbol))
        {
            throw new UnsupportedInputException(
                $"There is no double-struck Unicode letter for “{letter}”.");
        }
        _out.Text.Append(symbol);
    }

    private void RejectEnvironment()
    {
        var unit = NextUnit();
        string name = "";
        if (unit is not null)
        {
            try { name = Render(unit).Text.ToString(); }
            catch (Exception) { name = ""; }
        }
        if (name.Length == 0)
        {
            throw new UnsupportedInputException(
                "LaTeX environments need two-dimensional layout, which has no inline Unicode form.");
        }
        throw new UnsupportedInputException(
            $"The {name} environment needs two-dimensional layout, which has no inline Unicode form.");
    }

    // MARK: Helpers

    /// <summary>
    /// Replaces the ASCII hyphen with U+2212 MINUS SIGN. A hyphen is not a
    /// minus: it is shorter, sits lower, and reads as a word break. Only linear
    /// maths goes through here, so <c>\text{}</c> keeps its hyphens and
    /// "well-known" is not mangled.
    /// </summary>
    private static string Mathematical(string text) => text.Replace('-', '−');

    /// <summary>
    /// Wraps in parentheses unless it is a single character, so
    /// <c>\frac{1}{2}</c> gives ½ while <c>\frac{x+1}{y-2}</c> gives (x+1)∕(y−2).
    /// </summary>
    private static string Parenthesised(string text) =>
        Core.Text.CharacterCount(text) == 1 ? text : "(" + text + ")";

    private static bool ContainsScript(IReadOnlyList<Token> tokens) =>
        tokens.Any(token => token switch
        {
            Token.Sup or Token.Sub => true,
            Token.Group group => ContainsScript(group.Inner),
            _ => false,
        });
}
